### ISSUE TITLE
cgh.depends_on not working with memcpy events

### ISSUE THREAD

**Issue by jolly-chen** [AdaptiveCpp/AdaptiveCpp#1231]
**Bug summary**
I have a kernel with computation that depends on memcpy operations to multiple device vectors. I added a dependency using `cgh.depends_on(std::vector<sycl::event>)`, but this doesn't seem to have an effect because my computation kernel is still operating on uninitialized input values.

**To Reproduce**
```
#include <array>
#include <iostream>
#include <sycl/sycl.hpp>

int main() {
  sycl::queue queue{sycl::gpu_selector{}};
  constexpr int N = 10;
  std::vector<double> x(N, 1);
  std::vector<double> y(N, 10);
  double *a = sycl::malloc_device<double>(N, queue);
  double *b = sycl::malloc_device<double>(N, queue);

  std::vector<sycl::event> depends;
  depends.push_back(queue.memcpy(a, x.data(), x.size() * sizeof(double)));
  depends.push_back(queue.memcpy(b, y.data(), y.size() * sizeof(double)));

  queue.submit([&](sycl::handler &cgh) {
    cgh.depends_on(depends);
    cgh.single_task([=]() {
      printf("a:\n");
      for (int i = 0; i < N; i++) {
        printf("%f ", a[i]);
      }
      printf("\n");

      printf("b:\n");
      for (int i = 0; i < N; i++) {
        printf("%f ", b[i]);
      }
      printf("\n");
    });
  });

  return 0;
}
```

**Expected behavior**
The output should be:
```
a:
1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 
b:
10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 
```
What I get is:
```
a:
0.000000 1047852020.193862 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 
b:
0.000000 1047852020.193862 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 
```
However, if I add a `queue.wait()` or wait for each copy event individually before the printing kernel, I do get the expected result.

**Describe your setup**
* Built from source in release mode. Commit: c33e8c401f46122f2a5b028a46bbecadfb983612
* Describe the dependencies that AdaptiveCpp sits on top of in your setup: 
   * clang version and how clang was installed: 15, installed with APT
   * host compiler: c++
   * CUDA version: 11.8
* Compiled with CMake, targeting a NVIDIA GeForce RTX2070 GPU with `cuda:sm_75`

**Optional additional diagnostic information**
```
[;32m[AdaptiveCpp Info] [0mkernel_cache: Registering kernel ZZ4mainENKUlRN7hipsycl4sycl7handlerEE_clES2_EUlvE_
[;32m[AdaptiveCpp Info] [0mbackend_loader: Searching path for backend libs: '"/home/jollychen/adaptivecpp/lib/hipSYCL"'
[;32m[AdaptiveCpp Info] [0mbackend_loader: Successfully opened plugin: "/home/jollychen/adaptivecpp/lib/hipSYCL/librt-backend-ocl.so" for backend 'ocl'
[;32m[AdaptiveCpp Info] [0mbackend_loader: Successfully opened plugin: "/home/jollychen/adaptivecpp/lib/hipSYCL/librt-backend-cuda.so" for backend 'cuda'
[;32m[AdaptiveCpp Info] [0mbackend_loader: Successfully opened plugin: "/home/jollychen/adaptivecpp/lib/hipSYCL/librt-backend-omp.so" for backend 'omp'
[;32m[AdaptiveCpp Info] [0mRegistering backend: 'ocl'...
[;32m[AdaptiveCpp Info] [0mocl_hardware_manager: Hiding platform 'NVIDIA CUDA' because it has known incompatibilities. Set ACPP_RT_OCL_SHOW_ALL_DEVICES=1 to override.
[;32m[AdaptiveCpp Info] [0mocl_hardware_manager: Discovered OpenCL platform Intel(R) OpenCL
[;32m[AdaptiveCpp Info] [0mocl_hardware_manager: Hiding platform 'Intel(R) FPGA Emulation Platform for OpenCL(TM)' because it has known incompatibilities. Set ACPP_RT_OCL_SHOW_ALL_DEVICES=1 to override.
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Spawned for backend OpenCL with configuration: 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    memcpy lane: 0
[;32m[AdaptiveCpp Info] [0m    kernel lane: 1
[;32m[AdaptiveCpp Info] [0mRegistering backend: 'cuda'...
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Spawned for backend CUDA with configuration: 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    memcpy lane: 0
[;32m[AdaptiveCpp Info] [0m    memcpy lane: 1
[;32m[AdaptiveCpp Info] [0m    kernel lane: 2
[;32m[AdaptiveCpp Info] [0m    kernel lane: 3
[;32m[AdaptiveCpp Info] [0mRegistering backend: 'omp'...
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Spawned for backend OpenMP with configuration: 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    memcpy lane: 0
[;32m[AdaptiveCpp Info] [0m    kernel lane: 1
[;32m[AdaptiveCpp Info] [0mDiscovered devices from backend 'OpenCL': 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    vendor: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
[;32m[AdaptiveCpp Info] [0m    name: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
[;32m[AdaptiveCpp Info] [0mDiscovered devices from backend 'CUDA': 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    vendor: NVIDIA
[;32m[AdaptiveCpp Info] [0m    name: NVIDIA GeForce RTX 2070
[;32m[AdaptiveCpp Info] [0mDiscovered devices from backend 'OpenMP': 
[;32m[AdaptiveCpp Info] [0m  device 0: 
[;32m[AdaptiveCpp Info] [0m    vendor: the hipSYCL project
[;32m[AdaptiveCpp Info] [0m    name: hipSYCL OpenMP host device
[;32m[AdaptiveCpp Info] [0mdag_manager: DAG manager is alive!
[;32m[AdaptiveCpp Info] [0mruntime: ******* rt launch initiated ********
[;32m[AdaptiveCpp Info] [0mqueue: Constructed queue with node group id 1
[;32m[AdaptiveCpp Info] [0mdag_manager: Checking DAG flush opportunity...
[;32m[AdaptiveCpp Info] [0mdag_manager: Checking DAG flush opportunity...
[;32m[AdaptiveCpp Info] [0mdag_manager: Checking DAG flush opportunity...
[;32m[AdaptiveCpp Info] [0mruntime: ******* rt shutdown ********
[;32m[AdaptiveCpp Info] [0mdag_manager: Waiting for async worker...
[;32m[AdaptiveCpp Info] [0mdag_manager: Submitting asynchronous flush...
[;32m[AdaptiveCpp Info] [0mdag_builder: DAG contains operations: 
[;32m[AdaptiveCpp Info] [0m0. Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80} @node 0x55593d5dab20
[;32m[AdaptiveCpp Info] [0m1. Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80} @node 0x55593e230430
[;32m[AdaptiveCpp Info] [0m2. kernel: ZZ4mainENKUlRN7hipsycl4sycl7handlerEE_clES2_EUlvE_ @node 0x55593de31c90
[;32m[AdaptiveCpp Info] [0m    --> requires node @0x55593d5dab20 Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0m    --> requires node @0x55593e230430 Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0mdag_manager: waiting for async worker...
[;32m[AdaptiveCpp Info] [0mdag_manager [async]: Flushing!
[;32m[AdaptiveCpp Info] [0mdag_manager [async]: Submitting node to scheduler!
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Processing node 0x55593d5dab20 with 0 non-virtual requirement(s) and 0 direct requirement(s).
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Dispatching to lane 0: Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0minorder_executor: Processing node 0x55593d5dab20 with 0 non-virtual requirement(s) and 0 direct requirement(s).
[;32m[AdaptiveCpp Info] [0minorder_executor: Dispatching to lane 0x55593d7a7190: Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0mdag_manager [async]: Submitting node to scheduler!
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Processing node 0x55593e230430 with 0 non-virtual requirement(s) and 0 direct requirement(s).
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Dispatching to lane 1: Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0minorder_executor: Processing node 0x55593e230430 with 0 non-virtual requirement(s) and 0 direct requirement(s).
[;32m[AdaptiveCpp Info] [0minorder_executor: Dispatching to lane 0x55593dff17f0: Memcpy: CPU-Device0 #1 {0, 0, 0}+{1, 1, 80}-->CUDA-Device0 #1 {0, 0, 0}+{1, 1, 80}{1, 1, 80}
[;32m[AdaptiveCpp Info] [0mdag_manager [async]: Submitting node to scheduler!
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Processing node 0x55593de31c90 with 2 non-virtual requirement(s) and 2 direct requirement(s).
[;32m[AdaptiveCpp Info] [0mmulti_queue_executor: Dispatching to lane 2: kernel: ZZ4mainENKUlRN7hipsycl4sycl7handlerEE_clES2_EUlvE_
[;32m[AdaptiveCpp Info] [0minorder_executor: Processing node 0x55593de31c90 with 2 non-virtual requirement(s) and 2 direct requirement(s).
[;32m[AdaptiveCpp Info] [0m --> Synchronizes with other queue for node: 0x55593d5dab20
[;32m[AdaptiveCpp Info] [0m --> Synchronizes with other queue for node: 0x55593e230430
[;32m[AdaptiveCpp Info] [0minorder_executor: Dispatching to lane 0x55593dd3f310: kernel: ZZ4mainENKUlRN7hipsycl4sycl7handlerEE_clES2_EUlvE_
[;32m[AdaptiveCpp Info] [0mdag_manager [async]: DAG flush complete.
a:
0.000000 1047852020.193862 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 
b:
0.000000 1047852020.193862 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 10.000000 
[;32m[AdaptiveCpp Info] [0mdag_manager: Shutdown.
```

Thanks!

**Comment by illuhad** [comment]
Hi!

I suspect that this is not a dependency issue. The line `inorder_executor: Processing node 0x55593de31c90 with 2 non-virtual requirement(s) and 2 direct requirement(s).` in your log indicates that both of your dependencies are correctly taken into account.

Can you please try adding `queue.wait()` right before the `return 0` at the end of `main()`?

The SYCL specification explicitly states that the queue destructor should not block. It also does not require eager execution of kernels. I think what happens is that AdaptiveCpp by default queues up work in order to perform a batched submission.
In your case, it is only forced to begin execution at the end of `main` when the queue is destroyed. But then the `std::vector` objects containing the input data to copy from may already have been destroyed since they were declared after the queue, and might contain garbage after this point. Thus everything is executed correctly, but the input values are gone.
And since the queue processes work asynchronously without blocking in the destructor, the input values might also become invalid while the copy is in progress in the background.

So I believe that this program is UB.

tl;dr: Always ensure that there is at least one synchronization point for your work before the program exits :-)

**Comment by jolly-chen** [comment]
Wow, thanks for the quick reply! You're right, if I add a `queue.wait()` at the end, I do get the expected result. 
In my actual code I do have a synchronization point before exiting and copying back the result so it didn't occur to me that was the problem. But from your explanation, it's clear that I am running into the same issue there because I am queuing up a sequence of copyHtoD->compute operations with the same host buffer so by the time the actual memcpy happens the input values were already overwritten. Thank you for helping me figure that out!
